src/CorePolicy.cc

```diff
@@ -0,0 +1,247 @@
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <thread>
+#include "PerfUtils/TimeTrace.h"
+#include "PerfUtils/Util.h"
+#include "PerfUtils/Cycles.h"
+#include "CoreArbiter/CoreArbiterClient.h"
+
+#include "Arachne.h"
+#include "CorePolicy.h"
+#include "PerfStats.h"
+#include "CoreArbiter/ArbiterClientShim.h"
+
+using PerfUtils::Cycles;
+
+/*
+ * We will attempt to increase the number of cores if the idle core
+ * fraction (computed as idlePercentage * numSharedCores) is less than this
+ * number.
+ */
+const double maxIdleCoreFraction = 0.1;
+
+/*
+ * We will attempt to increase the number of cores if the load factor
+ * increases beyond this threshold.
+ */
+const double loadFactorThreshold = 1.0;
+
+/*
+ * Save the core fraction at which we ramped up based on load factor, so we
+ * can decide whether to ramp down.  Allocated in bootstrapLoadEstimator.
+ */
+double *utilizationThresholds = NULL;
+
+/*
+ * The difference in load, expressed as a fraction of a core, between a
+ * ramp-down threshold and the corresponding ramp-up threshold (i.e., we
+ * wait to ramp down until the load gets a bit below the point at 
+ * which we ramped up).
+ */
+const double idleCoreFractionHysteresis = 0.2;
+
+/*
+ * Do not ramp down if the percentage of occupied threadContext slots is
+ * above this threshold.
+ */
+const double SLOT_OCCUPANCY_THRESHOLD = 0.5;
+
+/*
+ * The period in ns over which we measure before deciding to reduce the number
+ * of cores we use.
+ */
+const uint64_t MEASUREMENT_PERIOD = 50 * 1000 * 1000;
+
+/* Is the core load estimator running? */
+bool loadEstimatorRunning = false;
+
+void coreLoadEstimator(CorePolicy* corePolicy);
+
+/*
+ * Choose a core to deschedule.  Return its coreId.
+ */
+int CorePolicy::chooseRemovableCore() {
+    CoreList* entry = threadClassCoreMap[defaultClass];
+    uint8_t minLoaded =
+        Arachne::occupiedAndCount[entry->map[0]]->load().numOccupied;
+    int minCoreId = entry->map[0];
+    for (uint32_t i = 1; i < threadClassCoreMap[defaultClass]->numFilled; i++) {
+        uint32_t coreId = entry->map[i];
+        if (Arachne::occupiedAndCount[coreId]->load().numOccupied < minLoaded) {
+            minLoaded = Arachne::occupiedAndCount[coreId]->load().numOccupied;
+            minCoreId = coreId;
+        }
+    }
+    return minCoreId;
+}
+
+/*
+ * Update threadClassCoreMap when a new core is added.  Callers must guarantee
+ * that this function is not called multiple times simultaneously, or at the
+ * same time as removeCore.

```

Comment:
Is this guarantee either necessary or sufficient? If not, I would eliminate it from the interface description (it might be simpler for CorePolicies to do their own synchronization than to clutter up the interface with  this requirement). Let's discuss.